feat: support nested event methods

andiwils opened the pull request:
This PR adds an optional path parameter to the `AddEvents` utility, enabling the specification of nested event locations within objects. The addition of the `getEvent` closure function allows for retrieving events based on this optional path, and an error is thrown if the event is not found.

A successful example:
```typescript
// Define a class with a nested event structure
class ExampleClass {
  nested = {
    onUpdate: new TypedEvent<() => void>()
  };
}

// Enhance the class with event handling capabilities specifying the path to the nested events
const EnhancedExampleClass = AddEvents<typeof ExampleClass, { onUpdate: TypedEvent<() => void> }>(ExampleClass, 'nested');

// Create an instance and attach an event listener
const exampleInstance = new EnhancedExampleClass();
exampleInstance.on('onUpdate', () => console.log('Event triggered'));

// Trigger the event
exampleInstance.nested.onUpdate.emit();
```

An error example:
```typescript

// Attempt to create an enhanced class with an incorrect path
try {
  const InvalidExampleClass = AddEvents<typeof ExampleClass, { onUpdate: TypedEvent<() => void> }>(ExampleClass, 'incorrectPath');
  const exampleInstance = new InvalidExampleClass();
  exampleInstance.on('onUpdate', () => console.log('This will not run'));
} catch (error) {
  console.error(error);  // Outputs: Error: Event "incorrectPath.onUpdate" is not defined
}
```

andiwils committed:
feat: support nested event methods

bbaldino commented on src/event-mixin.spec.ts:
```diff
@@ -76,3 +76,67 @@ describe('a child class', () => {
+
+  const NestedEventsClass = AddEvents<
+    typeof DeepNestedEvents,
+    {
+      eventOne: TypedEvent<(value: number) => void>;
+      eventTwo: TypedEvent<(value: boolean) => void>;
+    }
```
nit: for consistency, can you extract these out into an interface like the test above does? with them matching it makes it easier to see that what's different here is the path argument.

andiwils replied:
Sure thing. Done.

bbaldino commented on src/event-mixin.ts:
```diff
@@ -37,7 +72,7 @@ export function AddEvents<TBase extends Constructor, U>(Base: TBase) {
     on<K extends keyof U, E extends eventHandlerType<U[K]>>(eventName: K, handler: E) {
       // Even though we bypass type safety in the call (casting this as any), we've enforced it in the
       // method signature above, so it's still safe.
```
This comment looks like it isn't really relevant here anymore...maybe better as a note on `getEvent`?

andiwils replied:
Good catch. Done.

bbaldino reviewed (commented):
couple minor comments.  can you add a section to the README about this as well?

andiwils committed:
fix(tests): update tests for consistency

andiwils committed:
fix(tests): update comment location after refactor

andiwils committed:
chore(readme): add documentation about nested events

andiwils commented:
@bbaldino Thanks for the feedback. I fixed the tests to be more consistent, moved the stale comment into the comment `getEvent` block, and added a section in the readme.

andiwils committed:
chore: update readme

andiwils committed:
fix: update readme
